src/cuda/fusion/column_row_slicing.cu

```diff
@@ -0,0 +1,226 @@
+#include "../atomic.h"
+#include "../cuda_common.h"
+#include "../utils.h"
+#include "column_row_slicing.h"
+
+namespace gs {
+namespace impl {
+namespace fusion {
+
+inline __host__ __device__ int UpPower(int key) {
+  int ret = 1 << static_cast<uint32_t>(std::log2(key) + 1);
+  return ret;
+}
+
+__device__ inline uint32_t Hash32Shift(uint32_t key) {
+  key = ~key + (key << 15);  // # key = (key << 15) - key - 1;
+  key = key ^ (key >> 12);
+  key = key + (key << 2);
+  key = key ^ (key >> 4);
+  key = key * 2057;  // key = (key + (key << 3)) + (key << 11);
+  key = key ^ (key >> 16);
+  return key;
+}
+
+__device__ inline uint64_t Hash64Shift(uint64_t key) {
+  key = (~key) + (key << 21);  // key = (key << 21) - key - 1;
+  key = key ^ (key >> 24);
+  key = (key + (key << 3)) + (key << 8);  // key * 265
+  key = key ^ (key >> 14);
+  key = (key + (key << 2)) + (key << 4);  // key * 21
+  key = key ^ (key >> 28);
+  key = key + (key << 31);
+  return key;
+}
+
+/**
+ * @brief Used to judge whether a node is in a node set
+ *
+ * @tparam IdType
+ */
+template <typename IdType>
+struct NodeQueryHashmap {
```

Comment:
Seems more than one place use `NodeQueryHashmap`? Could it be moved to one shared header file?